<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no">
  <meta name="keywords" content="进进吖" />
  <meta name="description" content="进进吖的个人博客" />
  <meta http-equiv="x-dns-prefetch-control" content="on">
  <link rel="dns-prefetch" href="https://busuanzi.ibruce.info">
  <link rel="dns-prefetch" href="https://at.alicdn.com">
  
  <link rel="dns-prefetch" href="https://widget.daovoice.io">
  <link rel="dns-prefetch" href="https://widget-static-cdn.daovoice.io">
  <link rel="dns-prefetch" href="https://im.daovoice.io">
  
  
  
  <link rel="dns-prefetch" href="https://cdn.jsdelivr.net">
  <link rel="dns-prefetch" href="https://api.github.com">
  <link rel="dns-prefetch" href="https://avatars3.githubusercontent.com">
  
  <link rel="stylesheet" type="text/css" href="/./style/main.css">
	<link rel="shortcut icon" href="" title="Favicon">
  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.css">
	<title>promise的链式调用</title>
  
  
    <script>(function(i,s,o,g,r,a,m){i["DaoVoiceObject"]=r;i[r]=i[r]||function(){(i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;a.charset="utf-8";m.parentNode.insertBefore(a,m)})(window,document,"script",('https:' == document.location.protocol ? 'https:' : 'http:') + "//widget.daovoice.io/widget/7b3bba3c.js","daovoice");daovoice('init',{app_id: "7b3bba3c"});daovoice('update');
  </script>
  
<meta name="generator" content="Hexo 6.3.0"><link rel="alternate" href="/atom.xml" title="进进吖" type="application/atom+xml">
</head>
<body>
  <canvas id="pattern-placeholder" height="230"></canvas>
<div class="navbar-header">
  <a class="blog-title" href="/">进进吖</a>
  <a class="face-img" href="/">
    <img src="https://static.hufangyun.com/face.jpg">
  </a>
</div>
<main>
  <div class="article-title">
    
  
  <h1 class="title">
    promise的链式调用
  </h1>
  


    <ul class="article-info">
      <li>
        发布
        <time datetime="2020-11-25T08:30:42.697Z" itemprop="datePublished">2020-11-25</time>
      </li>
      <li>
        
    更新 <time datetime="2020-12-04T02:56:59.739Z" itemprop="dateUpdated">2020-12-04</time>

      </li>
      <li id="busuanzi_container_page_pv">
        阅读 <span id="busuanzi_value_page_pv"></span>
      </li>
    </ul>
  </div>
  <div class="container">
    <div class="article">
      <div class="content">
        
        <h3 id="promise链式调用"><a href="#promise链式调用" class="headerlink" title="promise链式调用"></a>promise链式调用</h3><ul>
<li><p>回调地狱</p>
</li>
<li><p>并发异步io操作,同一时间拿到结果</p>
</li>
<li><p>回调函数</p>
</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">fs.readFile(&#x27;./1.txt&#x27;,&#x27;utf8&#x27;,(err,data)=&gt;&#123;</span><br><span class="line">    fs.readFile(data,&#x27;utf8&#x27;,(err,data)=&gt;&#123;</span><br><span class="line">        fs.readFile(data,&#x27;utf8&#x27;,(err,data)=&gt;&#123;</span><br><span class="line">            console.log(data) // hello swr</span><br><span class="line">        &#125;)</span><br><span class="line">    &#125;)</span><br><span class="line">&#125;)</span><br></pre></td></tr></table></figure>

<ul>
<li>Promise<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br></pre></td><td class="code"><pre><span class="line">read(&#x27;./1.txt&#x27;,&#x27;utf8&#x27;)</span><br><span class="line">.then((data)=&gt;&#123;</span><br><span class="line">    // 1.如果一个promise执行完后，返回的还是一个promise，</span><br><span class="line">    //   会把这个promise的执行结果会传递给下一次then中</span><br><span class="line">    return read(data,&#x27;utf8&#x27;)</span><br><span class="line">&#125;)</span><br><span class="line">.then((data)=&gt;&#123;</span><br><span class="line">    return read(data,&#x27;utf8&#x27;)</span><br><span class="line">&#125;)</span><br><span class="line">.then((data)=&gt;&#123;</span><br><span class="line">    // 2.如果在then中返回的不是一个promise，</span><br><span class="line">    //   而是一个普通值，会将这个普通值作为下次then的成功的结果</span><br><span class="line">    return data.split(&#x27;&#x27;).reverse().join(&#x27;&#x27;)</span><br><span class="line">&#125;)</span><br><span class="line">.then((data)=&gt;&#123;</span><br><span class="line">    console.log(data) // rws olleh</span><br><span class="line">    // 3.如果当前then中失败了，会走下一个then的失败回调</span><br><span class="line">    throw new Error(&#x27;出错&#x27;)</span><br><span class="line">&#125;)</span><br><span class="line">.then(null,(err)=&gt;&#123;</span><br><span class="line">    console.log(err) // Error:出错   报错了</span><br><span class="line">    // 4.如果在then中不返回值，虽然没有显式返回，</span><br><span class="line">    //   但是默认是返回undefined，是属于普通值，依然会把这个普通值传到</span><br><span class="line">    //   下一个then的成功回调中</span><br><span class="line">&#125;)</span><br><span class="line">.then((data)=&gt;&#123;</span><br><span class="line">    console.log(data) // undefined</span><br><span class="line">&#125;)</span><br></pre></td></tr></table></figure>
<blockquote>
<p>1.如果一个promise执行完后，返回的还是一个promise，会把这个promise的执行结果会传递给下一次then中<br>2.如果在then中返回的不是一个promise，而是一个普通值，会将这个普通值作为下次then的成功的结果<br>3.如果当前then中失败了，会走下一个then的失败回调<br>4.如果在then中不返回值，虽然没有显式返回，但是默认是返回undefined，是属于普通值，依然会把这个普通值传到下一个then的成功回调中</p>
</blockquote>
</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br></pre></td><td class="code"><pre><span class="line">// 如果在then中抛出错误，会怎样呢？</span><br><span class="line">// 情景一，会被下一个then中的失败回调捕获</span><br><span class="line">read(&#x27;./1.txt&#x27;,&#x27;utf8&#x27;)</span><br><span class="line">.then((data)=&gt;&#123;</span><br><span class="line">    throw new Error(&#x27;出错了&#x27;)</span><br><span class="line">&#125;)</span><br><span class="line">.then(null,(err)=&gt;&#123;</span><br><span class="line">    console.log(err) // Error:出错了   报错</span><br><span class="line">&#125;)</span><br><span class="line"></span><br><span class="line">// 情景二，如果没有被失败的回调捕获，抛出错误最终会变成异常</span><br><span class="line">read(&#x27;./1.txt&#x27;,&#x27;utf8&#x27;)</span><br><span class="line">.then((data)=&gt;&#123;</span><br><span class="line">    throw new Error(&#x27;出错了&#x27;)</span><br><span class="line">&#125;)</span><br><span class="line"></span><br><span class="line">// 情景三，如果没有被失败的回调捕获，那么最终会被catch捕获到</span><br><span class="line">read(&#x27;./1.txt&#x27;,&#x27;utf8&#x27;)</span><br><span class="line">.then((data)=&gt;&#123;</span><br><span class="line">    throw new Error(&#x27;出错了&#x27;)</span><br><span class="line">&#125;)</span><br><span class="line">.then((data)=&gt;&#123;</span><br><span class="line">    </span><br><span class="line">&#125;)</span><br><span class="line">.catch((err)=&gt;&#123;</span><br><span class="line">    console.log(err) // Error:出错了   报错</span><br><span class="line">&#125;)</span><br><span class="line"></span><br><span class="line">// 情景四，如果被失败的回调捕获了，那么不会被catch捕获到</span><br><span class="line">read(&#x27;./1.txt&#x27;,&#x27;utf8&#x27;)</span><br><span class="line">.then((data)=&gt;&#123;</span><br><span class="line">    throw new Error(&#x27;出错了&#x27;)</span><br><span class="line">&#125;)</span><br><span class="line">.then(null,(err)=&gt;&#123;</span><br><span class="line">    console.log(err) // Error:出错了   报错</span><br><span class="line">&#125;)</span><br><span class="line">.catch((err)=&gt;&#123;</span><br><span class="line">    console.log(err)  // 不会执行到这里</span><br><span class="line">&#125;)</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<blockquote>
<p>then 里面可以不写东西<br>catch是错误没有处理的情况下才能执行</p>
</blockquote>
<h3 id="实现promise链式调用"><a href="#实现promise链式调用" class="headerlink" title="实现promise链式调用"></a>实现promise链式调用</h3><figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">let promise = new Promise((resolve,reject)=&gt;&#123;</span><br><span class="line">    resolve()</span><br><span class="line">&#125;)</span><br><span class="line"></span><br><span class="line">// 2.返回的值为promise2,这是promiseA+规范规定的</span><br><span class="line">let promise2 = promise.then((data)=&gt;&#123;</span><br><span class="line">    return x // then中的返回值x可能是普通值也可能是promise，并且传给下一个then</span><br><span class="line">&#125;).then((data)=&gt;&#123;</span><br><span class="line">    console.log(data) // x的值</span><br><span class="line">&#125;)</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br></pre></td><td class="code"><pre><span class="line">// 源码</span><br><span class="line">function Promise(executor)&#123;</span><br><span class="line">    let self = this</span><br><span class="line">    self.value = undefined  // 成功的值</span><br><span class="line">    self.reason = undefined // 失败的值</span><br><span class="line">    self.status = &#x27;pending&#x27; // 初始化状态</span><br><span class="line">    self.onResolvedCallbacks = []  // 成功的回调</span><br><span class="line">    self.onRejectedCallbacks = []  // 失败的回调</span><br><span class="line">    function resolve(value)&#123;</span><br><span class="line">        if(self.status === &#x27;pending&#x27;)&#123;  // 只有在pending的状态才能转为成功态</span><br><span class="line">            self.value = value </span><br><span class="line">            self.status = &#x27;resolved&#x27;</span><br><span class="line">            self.onResolvedCallbacks.forEach(fn=&gt;fn())  // 可能new Promise的时候会存在异步操作，把成功和失败的回调保存起来</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">    function reject(reason)&#123;</span><br><span class="line">        if(self.status === &#x27;pending&#x27;)&#123; // 只有在pending的状态才能转为失败态</span><br><span class="line">            self.reason = reason </span><br><span class="line">            self.status = &#x27;rejected&#x27;</span><br><span class="line">            self.onRejectedCallbacks.forEach(fn=&gt;fn())</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">    try&#123;</span><br><span class="line">         // 在new Promise的时候，立即执行的函数，称为执行器</span><br><span class="line">        executor(resolve,reject)</span><br><span class="line">    &#125;catch(error)&#123; // 如果执行executor抛出错误，则会走失败reject</span><br><span class="line">        reject(error)</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">Promise.prototype.then = function(onFulfilled,onRejected)&#123;</span><br><span class="line">    let self = this</span><br><span class="line">    let promise2 // promise链式调用时，返回的是一个新的promise对象，在这里声明一个新的promise</span><br><span class="line">    </span><br><span class="line">     promise2 = new Promise((resolve,reject)=&gt;&#123;</span><br><span class="line">        if(self.status === &#x27;resolved&#x27;)&#123;</span><br><span class="line">            // 当执行成功回调的时候，可能会出现异常，那么就把这个异常作为promise2的错误的结果</span><br><span class="line">            try&#123;</span><br><span class="line">                let x = onFulfilled(self.value) // x，代表then中执行完返回的结果</span><br><span class="line">                // 根据promiseA+规范，我们应该提供一个函数来处理promise2</span><br><span class="line">                //   个人的理解:then中不管是成功回调还是失败回调，其返回</span><br><span class="line">                //   值，有可能是promise，也有可能是普通值，也有可能是抛出错误</span><br><span class="line">                //   那么就需要一个函数来处理这几种不同的情况</span><br><span class="line">                resolvePromise(promise2,x,resolve,reject)</span><br><span class="line">                // 这里的promise2就是当前的promise2，x则是执行then中成功回调后返回的结果，如果是成功则调promise2的resolve，失败则调reject</span><br><span class="line">            &#125;catch(e)&#123;</span><br><span class="line">                reject(e) // 注意：这里的reject是这个promise2的reject</span><br><span class="line">            &#125;</span><br><span class="line">        &#125;</span><br><span class="line">        if(self.status === &#x27;rejected&#x27;)&#123;</span><br><span class="line">            try&#123;</span><br><span class="line">                let x = onRejected(self.reason) </span><br><span class="line">                resolvePromise(promise2,x,resolve,reject)</span><br><span class="line">            &#125;catch(e)&#123;</span><br><span class="line">                reject(e)</span><br><span class="line">            &#125;</span><br><span class="line">        &#125;</span><br><span class="line">        if(self.status === &#x27;pending&#x27;)&#123;</span><br><span class="line">            self.onResolvedCallbacks.push(()=&gt;&#123;</span><br><span class="line">                try&#123;</span><br><span class="line">                    let x =  onFulfilled(self.value)</span><br><span class="line">                    resolvePromise(promise2,x,resolve,reject)</span><br><span class="line">                &#125;catch(e)&#123;</span><br><span class="line">                    reject(e)</span><br><span class="line">                &#125;</span><br><span class="line">            &#125;) </span><br><span class="line">            self.onRejectedCallbacks.push(()=&gt;&#123;</span><br><span class="line">                try&#123;</span><br><span class="line">                    let x = onRejected(self.reason)</span><br><span class="line">                    resolvePromise(promise2,x,resolve,reject)</span><br><span class="line">                &#125;catch(e)&#123;</span><br><span class="line">                    reject(e)</span><br><span class="line">                &#125;</span><br><span class="line">            &#125;)</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;)</span><br><span class="line">    return promise2 // 在jquery中是return this，但是在promise中，则是返回一个新的promise对象</span><br><span class="line">&#125;</span><br><span class="line">module.exports = Promise</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<h3 id="完整代码"><a href="#完整代码" class="headerlink" title="完整代码"></a>完整代码</h3><figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br></pre></td><td class="code"><pre><span class="line">function Promise(executor) &#123;</span><br><span class="line">    let self = this;</span><br><span class="line">    self.value = undefined;  // 成功的值</span><br><span class="line">    self.reason = undefined;  // 失败的值</span><br><span class="line">    self.status = &#x27;pending&#x27;; // 目前promise的状态pending</span><br><span class="line">    self.onResolvedCallbacks = []; // 可能new Promise的时候会存在异步操作，把成功和失败的回调保存起来</span><br><span class="line">    self.onRejectedCallbacks = [];</span><br><span class="line">    function resolve(value) &#123; // 把状态更改为成功</span><br><span class="line">        if (self.status === &#x27;pending&#x27;) &#123; // 只有在pending的状态才能转为成功态</span><br><span class="line">            self.value = value;</span><br><span class="line">            self.status = &#x27;resolved&#x27;;</span><br><span class="line">            self.onResolvedCallbacks.forEach(fn =&gt; fn()); // 把new Promise时异步操作，存在的成功回调保存起来</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">    function reject(reason) &#123;  // 把状态更改为失败</span><br><span class="line">        if (self.status === &#x27;pending&#x27;) &#123; // 只有在pending的状态才能转为失败态</span><br><span class="line">            self.reason = reason;</span><br><span class="line">            self.status = &#x27;rejected&#x27;;</span><br><span class="line">            self.onRejectedCallbacks.forEach(fn =&gt; fn()); // 把new Promise时异步操作，存在的失败回调保存起来</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">    try &#123;</span><br><span class="line">        // 在new Promise的时候，立即执行的函数，称为执行器</span><br><span class="line">        executor(resolve, reject);</span><br><span class="line">    &#125; catch (e) &#123; // 如果执行executor抛出错误，则会走失败reject</span><br><span class="line">        reject(e);</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br></pre></td><td class="code"><pre><span class="line">Promise.prototype.then = function(onFulfilled,onRejected)&#123;</span><br><span class="line">    // onFulfilled、onRejected是可选参数</span><br><span class="line">    onFulfilled = typeof onFulfilled === &#x27;function&#x27;?onFulfilled:val=&gt;val;</span><br><span class="line">    onRejected = typeof onRejected === &#x27;function&#x27;?onRejected: err=&gt;&#123;throw err&#125;;</span><br><span class="line">    let self = this;</span><br><span class="line">    let promise2; </span><br><span class="line">     // promise和jquery的区别是 promise不能单纯返回自身，</span><br><span class="line">    // 而是每次都是返回一个新的promise，才可以实现链式调用，</span><br><span class="line">   // 因为同一个promise的pending resolve reject只能更改一次</span><br><span class="line">     promise2 = new Promise((resolve,reject)=&gt;&#123;</span><br><span class="line">        if(self.status === &#x27;resolved&#x27;)&#123;</span><br><span class="line">          // 为什么要加setTimeout？</span><br><span class="line">         // 首先是promiseA+规范要求的</span><br><span class="line">        // 其次是代码，有的是同步，有的是异步</span><br><span class="line">       // 所以为了更加统一，就使用为setTimeout变为异步了，保持一致性</span><br><span class="line">            setTimeout(()=&gt;&#123;</span><br><span class="line">              try &#123; // 上面executor虽然使用try catch捕捉错误</span><br><span class="line">                // 但是在异步中，不一定能够捕捉，所以在这里</span><br><span class="line">                // 用try catch捕捉</span><br><span class="line">              let x = onFulfilled(self.value);</span><br><span class="line">              // 在then中，返回值可能是一个promise，所以</span><br><span class="line">              // 需要resolvePromise对返回值进行判断</span><br><span class="line">               resolvePromise(promise2,x,resolve,reject)</span><br><span class="line">                &#125;catch(e)&#123;</span><br><span class="line">                    reject(e) </span><br><span class="line">                &#125;</span><br><span class="line">            &#125;,0)</span><br><span class="line">        &#125;</span><br><span class="line">        </span><br><span class="line">        if(self.status === &#x27;rejected&#x27;)&#123;</span><br><span class="line">            setTimeout(()=&gt;&#123;</span><br><span class="line">                try&#123;</span><br><span class="line">                    let x = onRejected(self.reason) </span><br><span class="line">                    resolvePromise(promise2,x,resolve,reject)</span><br><span class="line">                &#125;catch(e)&#123;</span><br><span class="line">                    reject(e)</span><br><span class="line">                &#125;</span><br><span class="line">            &#125;,0)</span><br><span class="line">        &#125;</span><br><span class="line">       // 当new Promise中有resolve、reject处于异步中，执行then的时候，状态为pending，</span><br><span class="line">        if(self.status === &#x27;pending&#x27;)&#123;</span><br><span class="line">            // 分别把成功、失败的回调函数，存到该数组中，这样写的好处，就是把参数传进去，不需要将来遍历onResolvedCallbacks、onRejectedCallbacks时，再传参</span><br><span class="line">            self.onResolvedCallbacks.push(()=&gt;&#123;</span><br><span class="line">                setTimeout(()=&gt;&#123;</span><br><span class="line">                    try&#123;</span><br><span class="line">                        let x =  onFulfilled(self.value)</span><br><span class="line">                        resolvePromise(promise2,x,resolve,reject)</span><br><span class="line">                    &#125;catch(e)&#123;</span><br><span class="line">                        reject(e)</span><br><span class="line">                    &#125;</span><br><span class="line">                &#125;,0)</span><br><span class="line">            &#125;) </span><br><span class="line"></span><br><span class="line">            self.onRejectedCallbacks.push(()=&gt;&#123;</span><br><span class="line">                setTimeout(()=&gt;&#123;</span><br><span class="line">                    try&#123;</span><br><span class="line">                        let x = onRejected(self.reason)</span><br><span class="line">                        resolvePromise(promise2,x,resolve,reject)</span><br><span class="line">                    &#125;catch(e)&#123;</span><br><span class="line">                        reject(e)</span><br><span class="line">                    &#125;</span><br><span class="line">                &#125;,0)</span><br><span class="line">            &#125;)</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;)</span><br><span class="line">    return promise2</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>


<p><strong>说明 call 的意义</strong></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">promise.then(()=&gt;&#123;</span><br><span class="line">    console.log(this) // 此时this是指向该promise的，对象的方法中this是指向这个对象的</span><br><span class="line">&#125;)</span><br></pre></td></tr></table></figure>
<blockquote>
<p>但是我们在下面通过let then &#x3D; promise.then，来判断是否promise，是否会异常<br>当我们执行then时，里面的this还是会指向这个promise吗？答案是不一定的，<br>因为此时then，如果在全局下执行，指向的可能就是window了，所以为了让this的<br>指向正确，需要通过<br>then.call(promise)，来把then的this指向promise</p>
</blockquote>
<h3 id="这个函数为核心主要是处理then中返回的值x和promise2的关系"><a href="#这个函数为核心主要是处理then中返回的值x和promise2的关系" class="headerlink" title="这个函数为核心主要是处理then中返回的值x和promise2的关系"></a>这个函数为核心主要是处理then中返回的值x和promise2的关系</h3><figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br></pre></td><td class="code"><pre><span class="line">function resolvePromise(promise2,x,resolve,reject)&#123;</span><br><span class="line">// 当promise2和then返回的值x为同一个对象时，变成了自己等自己，会陷入死循环</span><br><span class="line">if(promise2 === x)&#123;</span><br><span class="line">  return reject(new TypeError(&#x27;Chaining cycle&#x27;));</span><br><span class="line">&#125;</span><br><span class="line">let called;</span><br><span class="line">// x可能是一个promise也可能是一个普通值</span><br><span class="line">if(x!==null &amp;&amp; (typeof x=== &#x27;object&#x27; || typeof x === &#x27;function&#x27;))&#123;</span><br><span class="line">  try&#123;</span><br><span class="line">      // 这里的x，就是上面then中执行完返回的结果，我们在这里声明一个x用来接收</span><br><span class="line">      // 返回的必须是一个新的promise，因为promise成功后不能再走失败</span><br><span class="line">      let then = x.then; </span><br><span class="line">      // x可能还是个promise，那么就让这个promise执行</span><br><span class="line">      // 但是还是存在一个恶作剧的情况，就是&#123;then:&#123;&#125;&#125;</span><br><span class="line">      // 此时需要新增一个判断then是否函数</span><br><span class="line">      if(typeof then === &#x27;function&#x27;)&#123;</span><br><span class="line">        then.call(x,y=&gt;&#123;  // y 是返回promise后的成功结果</span><br><span class="line">              if(called) return; </span><br><span class="line">              called = true;</span><br><span class="line">      // 只能创建一个新的promise再执行业务逻辑，返回同一个promise的话，就不能既成功又失败</span><br><span class="line">              resolvePromise(promise2,y,resolve,reject);</span><br><span class="line">          &#125;,err=&gt;&#123; </span><br><span class="line">              if(called) return;</span><br><span class="line">              called = true;</span><br><span class="line">              reject(err);</span><br><span class="line">          &#125;);</span><br><span class="line">      &#125;else&#123;</span><br><span class="line">          resolve(x);</span><br><span class="line">      &#125;</span><br><span class="line">  &#125;catch(e)&#123;</span><br><span class="line">      if(called) return;</span><br><span class="line">      called = true;</span><br><span class="line">      reject(e);</span><br><span class="line">  &#125;</span><br><span class="line"> &#125;else&#123; </span><br><span class="line">  resolve(x);</span><br><span class="line"> &#125;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
      </div>
        <div class="support-author">
          <p>感谢您的阅读。 🙏
          <a href="" target="_blank">关于转载请看这里</a>
            <!--<a class="btn-pay"  href="#pay-modal">¥ 打赏支持</a>-->
          </p>
        </div>
        <!--
            <div class="like ">
              <div class="like-button">
                <a id="like-note" href="">
                  <i class="icon-heart"></i>喜欢
                </a>
              </div>
              <span id="likes-count">256</span>
            </div>
        -->
        <div class="otherLink">
          <div class="previous">
          </div>
          <div class="next">
          </div>
        </div>
        <div class="comments" id="comments">
          
<script src="https://cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.js"></script>
<script type="text/javascript">
  const gitalk = new Gitalk({
    clientID: 'd0c6de133269f4b5e478',
    clientSecret: '553bce00ff56a439112749cdde0939a6ebc77d32',
    repo: 'https://zhangzengjin1213.github.io/',
    owner: 'zhangzengjin1213',
    admin: ['zhangzengjin1213'],
    id: location.pathname,      // Ensure uniqueness and length less than 50
    distractionFreeMode: false
  })

  gitalk.render('comments');
</script>


        </div>
      </div>
    </div>
   </div>
</main>
<div class="footer">
  <div class="info">
    <p>
    <a target="_blank" rel="noopener" href="https://hexo.io"> Hexo </a> 强力驱动 |
      <a target="_blank" rel="noopener" href="https://github.com/Youthink/hexo-themes-yearn"> Yearn </a>
      主题
    </p>
    <p>&copy;2019-04 进进的博客 京ICP备1314520号</p>
  </div>
</div>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<script>//console
  var consoleConfig = '\n欢迎访问 https://hufangyun.com ，围观小猿大圣的博客(づ｡◕‿‿◕｡)づ！\n,\n本博客使用 %cHexo%c 搭建，博客主题为小猿大圣开发的 %chexo-themes-yearn%c ~~~ 🎉🎉🎉 \n\n源码 https://github.com/Youthink/hexo-themes-yearn \n\n如果喜欢可以 star 支持一下 ❤️~\n,\n 想知道这个效果如何实现的？博客内搜索 console 彩蛋 🚀 ！\n'.split(',');
  var canConsole = true;
  var consoleInfo = (function(consoleConfig) {
  if (!canConsole || !consoleConfig || consoleConfig.length < 1) {
    return;
  }
  var consoleColor = '#6190e8';
  var _console;
  var backgroundTextStyle = 'padding: 1px 5px;color: #fff;background: ' + consoleColor + ';'
  var textStyle = 'color: ' + consoleColor + ';';

  consoleConfig.map(o => {
    var num = (o.match(/%c/g) || []).length;
    if(/^http(s)?:\/\//.test(o)) {
      console.log('%c     ', 'background: url(' + o + ') no-repeat left center;font-size: 180px;');
      return;
    }
    if (num > 0) {
      var logArguments = [];
      for (var i = 0; i < num; i++) {
        if (i % 2 === 0) {
          logArguments.push(backgroundTextStyle);
        } else {
          logArguments.push(textStyle);
        }
      }
      (_console = console).log.apply(_console, ['%c' + o, textStyle].concat(logArguments));
      return;
    }
    console.log('%c' + o, textStyle);
  });
}(consoleConfig));</script><script type="text/javascript" src="/./js/main.js"></script>

  <script src="//at.alicdn.com/t/font_159214_mvtxvg9me9.js"></script>
</body>
</html>
